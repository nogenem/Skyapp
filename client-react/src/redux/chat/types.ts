import { TMessageType } from '~/constants/message_types';
import { TUserStatus } from '~/constants/user_status';

interface IOtherUser {
  _id: string;
  nickname: string;
  thoughts: string;
  status: TUserStatus;
  online: boolean;
  channel_id?: string;
}

interface IOtherUsers {
  [_id: string]: IOtherUser;
}

interface IMember {
  user_id: string;
  is_adm: boolean;
  last_seen: Date;
}

interface IChannel {
  _id: string;
  name: string;
  is_group: boolean;
  members: IMember[];
  other_member_idx?: number;
  unread_msgs: number;
  lastMessage?: IMessage;
}

interface IChannels {
  [_id: string]: IChannel;
}

interface IAttachment {
  originalName: string;
  size: number;
  path: string;
  mimeType: string;
  imageDimensions?: { width: number; height: number };
}

interface IMessage {
  _id: string;
  channel_id: string;
  from_id?: string;
  body: string | IAttachment;
  type: TMessageType;
  createdAt: Date;
  updatedAt: Date;
}

interface IInitialData {
  users: IOtherUsers;
  channels: IChannels;
}

interface IActiveChannelInfo {
  _id: string;
  messages: IMessage[];
  totalMessages: number;
  queue: IMessage[];
}

interface INewGroupCredentials {
  name: string;
  members: string[];
  admins: string[];
}

interface IUpdateGroupCredentials {
  channel_id: string;
  name: string;
  members: string[];
  admins: string[];
}

interface ILeaveGroupCredentials {
  channel_id: string;
}

interface IFetchMessagesCredentials {
  channel_id: string;
  offset: number;
  limit?: number;
  sort?: string;
}

interface ISendMessageCredentials {
  channel_id: string;
  body: string;
}

type TChatState = {
  users: IOtherUsers;
  channels: IChannels;
  activeChannelInfo?: IActiveChannelInfo;
};

enum EChatActions {
  SET_INITIAL_DATA = '@chat/SET_INITIAL_DATA',
  SET_USER_ONLINE = '@chat/SET_USER_ONLINE',
  ADD_OR_UPDATE_CHANNEL = '@chat/ADD_OR_UPDATE_CHANNEL',
  SET_ACTIVE_CHANNEL = '@chat/SET_ACTIVE_CHANNEL',
  REMOVE_CHANNEL = '@chat/REMOVE_CHANNEL',
  ADD_NEW_USER = '@chat/ADD_NEW_USER',
  ADD_MESSAGES = '@chat/ADD_MESSAGES',
  SET_LATEST_MESSAGE = '@chat/SET_LATEST_MESSAGE',
  ADD_TO_MESSAGES_QUEUE = '@chat/ADD_TO_MESSAGES_QUEUE',
  REMOVE_FROM_MESSAGES_QUEUE = '@chat/REMOVE_FROM_MESSAGES_QUEUE',
  SET_LAST_SEEN = '@chat/SET_LAST_SEEN',
  SET_USER_STATUS = '@chat/SET_USER_STATUS',
  SET_USER_THOUGHTS = '@chat/SET_USER_THOUGHTS',
}

interface IChatActionType<T, P> {
  type: T;
  payload: P;
}

type TChatAction =
  | IChatActionType<typeof EChatActions.SET_INITIAL_DATA, IInitialData>
  | IChatActionType<
      typeof EChatActions.SET_USER_ONLINE,
      { _id: string; value: boolean }
    >
  | IChatActionType<typeof EChatActions.ADD_OR_UPDATE_CHANNEL, IChannel>
  | IChatActionType<
      typeof EChatActions.SET_ACTIVE_CHANNEL,
      { _id: string | undefined }
    >
  | IChatActionType<typeof EChatActions.REMOVE_CHANNEL, { channelId: string }>
  | IChatActionType<typeof EChatActions.ADD_NEW_USER, IOtherUser>
  | IChatActionType<
      typeof EChatActions.ADD_MESSAGES,
      { messages: IMessage[]; totalMessages: number; atTop: boolean }
    >
  | IChatActionType<typeof EChatActions.SET_LATEST_MESSAGE, IMessage>
  | IChatActionType<typeof EChatActions.ADD_TO_MESSAGES_QUEUE, IMessage>
  | IChatActionType<typeof EChatActions.REMOVE_FROM_MESSAGES_QUEUE, IMessage>
  | IChatActionType<
      typeof EChatActions.SET_LAST_SEEN,
      {
        channel_id: string;
        user_id: string;
        last_seen: string;
      }
    >
  | IChatActionType<
      typeof EChatActions.SET_USER_STATUS,
      {
        user_id: string;
        newStatus: TUserStatus;
      }
    >
  | IChatActionType<
      typeof EChatActions.SET_USER_THOUGHTS,
      {
        user_id: string;
        newThoughts: string;
      }
    >;

export type {
  TChatState,
  TChatAction,
  IInitialData,
  IOtherUsers,
  IOtherUser,
  IChannel,
  IChannels,
  IMember,
  IMessage,
  IAttachment,
  INewGroupCredentials,
  IUpdateGroupCredentials,
  ILeaveGroupCredentials,
  IFetchMessagesCredentials,
  ISendMessageCredentials,
};
export { EChatActions };
