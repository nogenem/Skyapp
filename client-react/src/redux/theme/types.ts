type TThemeMode = 'dark' | 'light';

type TThemeState = {
  mode?: TThemeMode;
};

enum EThemeActions {
  MODE_SWITCHED = '@theme/MODE_SWITCHED',
}

interface IThemeActionType<T, P> {
  type: T;
  payload: P;
}

type TThemeAction = IThemeActionType<
  typeof EThemeActions.MODE_SWITCHED,
  TThemeMode
>;

export type { TThemeMode, TThemeState, TThemeAction };
export { EThemeActions };
