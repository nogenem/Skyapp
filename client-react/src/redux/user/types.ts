interface User {
  _id: string;
  token?: string;
}

type UserState = User;

enum EUserActions {
  SIGNED_IN = '@user/SIGNED_IN',
  SIGNED_OUT = '@user/SIGNED_OUT',
}

interface UserActionType<T, P> {
  type: T;
  payload: P;
}

type UserAction =
  | UserActionType<typeof EUserActions.SIGNED_IN, User>
  | UserActionType<typeof EUserActions.SIGNED_OUT, null>;

export type { User, UserState, UserAction };
export { EUserActions };
