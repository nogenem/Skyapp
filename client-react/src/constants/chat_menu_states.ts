export const MENU_STATES = {
  MAIN: 1,
  CHANGING_LANGUAGE: 2,
  CHANGING_USER_STATUS: 3,
  CHANGING_USER_THOUGHTS: 4,
} as const;
type TMenuStates = typeof MENU_STATES[keyof typeof MENU_STATES];

export type { TMenuStates };
