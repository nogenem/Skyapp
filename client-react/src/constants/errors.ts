export const INVALID_EMAIL = "Invalid email";
export const CANT_BE_BLANK = "Can't be blank";
export const PASSWORDS_MUST_MATCH = "Passwords must match";
export const fieldIsTooShort = (n: number) =>
  `This field must be at least ${n} chars long`;
