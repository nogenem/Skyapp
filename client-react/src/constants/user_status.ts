export const USER_STATUS = {
  ACTIVE: 1,
  AWAY: 2,
  DO_NOT_DISTURB: 3,
  INVISIBLE: 4,
} as const;

export const USER_STATUS_2_TEXT = {
  [USER_STATUS.ACTIVE]: 'Active',
  [USER_STATUS.AWAY]: 'Away',
  [USER_STATUS.DO_NOT_DISTURB]: 'Do not disturb',
  [USER_STATUS.INVISIBLE]: 'Invisible',
} as const;

type TUserStatus = typeof USER_STATUS[keyof typeof USER_STATUS];

export type { TUserStatus };
