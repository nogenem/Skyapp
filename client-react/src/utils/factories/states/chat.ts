import faker from 'faker/locale/en_US';

import { initialState } from '~/redux/chat/reducer';
import type {
  IChannel,
  IChannels,
  IOtherUser,
  IOtherUsers,
  TChatState,
} from '~/redux/chat/types';

import activeChannelInfoFactory from '../models/activeChannelInfo';
import channelFactory from '../models/channel';
import otherUserFactory from '../models/otherUser';

interface IOptions {
  useConstValues: boolean;
  useInitialState: boolean;
  usersLen: number;
  channelsLen: number;
}

export default (
  override?: Partial<TChatState>,
  options?: Partial<IOptions>,
): TChatState => {
  let usersLen = faker.datatype.number({ min: 0, max: 2 });
  if (options?.usersLen !== undefined && options.usersLen >= 0) {
    usersLen = options.usersLen;
  } else if (options?.useConstValues) {
    usersLen = 2;
  }

  let tmpOtherUsers: IOtherUser[] = [];
  if (!options?.useConstValues) {
    tmpOtherUsers = Array.from({ length: usersLen }, () => otherUserFactory());
  } else {
    tmpOtherUsers = Array.from({ length: usersLen }, (_, idx) =>
      otherUserFactory(
        { _id: `other-user-${idx}`, channel_id: `channel-${idx}` },
        { useConstValues: true },
      ),
    );
  }

  const users: IOtherUsers = {};
  tmpOtherUsers.forEach(user => {
    users[user._id] = user;
  });

  let channelsLen = faker.datatype.number({ min: 0, max: 2 });
  if (options?.channelsLen !== undefined && options.channelsLen >= 0) {
    channelsLen = options.channelsLen;
  } else if (options?.useConstValues) {
    channelsLen = 2;
  }

  let tmpChannels: IChannel[] = [];
  if (!options?.useConstValues) {
    tmpChannels = Array.from({ length: channelsLen }, () => channelFactory());
  } else {
    tmpChannels = Array.from({ length: channelsLen }, (_, idx) =>
      channelFactory({ _id: `channel-${idx}` }, { useConstValues: true }),
    );
  }

  const channels: IChannels = {};
  tmpChannels.forEach(channel => {
    channels[channel._id] = channel;
  });

  let activeChannelInfo =
    Math.random() < 0.5 ? undefined : activeChannelInfoFactory();
  if (options?.useConstValues) activeChannelInfo = undefined;

  return options?.useInitialState
    ? initialState
    : ({
        users,
        channels,
        activeChannelInfo,
        ...(override || {}),
      } as TChatState);
};

export type { IOptions };
