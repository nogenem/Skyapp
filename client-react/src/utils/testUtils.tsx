import React from 'react';
import { Provider } from 'react-redux';

import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import type { IAppState } from '~/redux/store';

export const getMockStore = () => {
  const middlewares = [thunk];
  return configureStore(middlewares);
};

export const getRenderWithRedux = () => {
  const emptyState: Partial<IAppState> = {};
  const mockStore = getMockStore();

  return (
    ui: React.ReactNode,
    initialState: Partial<IAppState> = emptyState,
  ) => {
    const store = mockStore(initialState);
    return {
      ...render(<Provider store={store}>{ui}</Provider>),
    };
  };
};
