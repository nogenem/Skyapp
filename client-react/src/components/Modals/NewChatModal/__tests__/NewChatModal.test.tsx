import React from 'react';

import { fireEvent, waitFor } from '@testing-library/react';

import { USER_STATUS } from '~/constants/user_status';
import { IOtherUser } from '~/redux/chat/types';
import { getRenderWithRedux } from '~/utils/testUtils';

import {
  NewChatModal as ConnectedNewChatModal,
  UnconnectedNewChatModal,
} from '../index';

const renderWithRedux = getRenderWithRedux();

describe('Connected NewChatModal', () => {
  it('renders correctly', () => {
    const otherUser: IOtherUser = {
      _id: '123456',
      nickname: 'Test',
      thoughts: '',
      status: USER_STATUS.ACTIVE,
      online: true,
    };
    const { getByRole } = renderWithRedux(
      <ConnectedNewChatModal isOpen onClose={() => {}} />,
      {
        chat: {
          users: {
            [otherUser._id]: otherUser,
          },
          channels: {},
        },
      },
    );

    expect(getByRole('presentation')).toMatchSnapshot();
  });
});

describe('Unconnected NewChatModal', () => {
  it('handles start chatting with other people', async () => {
    const otherUser: IOtherUser = {
      _id: '123456',
      nickname: 'Test',
      thoughts: '',
      status: USER_STATUS.ACTIVE,
      online: true,
    };
    const users = [otherUser];
    const onClose = jest.fn(() => {});
    const sendCreateChannelWith = jest.fn(() => Promise.resolve());

    const { getByText } = renderWithRedux(
      <UnconnectedNewChatModal
        isOpen
        onClose={onClose}
        users={users}
        sendCreateChannelWith={sendCreateChannelWith}
      />,
    );

    expect(getByText(otherUser.nickname)).toBeInTheDocument();

    fireEvent.click(getByText(otherUser.nickname));

    await waitFor(() => expect(onClose).toHaveBeenCalled());

    expect(sendCreateChannelWith).toHaveBeenCalled();
    expect(onClose).toHaveBeenCalled();
  });
});
