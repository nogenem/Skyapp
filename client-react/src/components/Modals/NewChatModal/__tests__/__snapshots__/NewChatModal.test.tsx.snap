// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connected NewChatModal renders correctly 1`] = `
<div
  class="MuiDialog-root"
  role="presentation"
  style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-test="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiDialog-container MuiDialog-scrollPaper"
    role="none presentation"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    tabindex="-1"
  >
    <div
      aria-labelledby="new-chat-modal-title"
      class="MuiPaper-root MuiDialog-paper makeStyles-dialogPaper-1 MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth MuiPaper-elevation24 MuiPaper-rounded"
      role="dialog"
    >
      <div
        class="MuiDialogTitle-root"
        id="new-chat-modal-title"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6"
        >
          Messages:Chat with someone
        </h2>
      </div>
      <div
        class="MuiDialogContent-root MuiDialogContent-dividers"
      >
        <div
          class="MuiFormControl-root MuiTextField-root makeStyles-textInput-4 MuiFormControl-marginNormal MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            Common:Search
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              name="search"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-7"
              >
                <span>
                  Common:Search
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <ul
          class="MuiList-root MuiList-padding"
        >
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemAvatar-root makeStyles-listAvatar-2"
            >
              <div
                class="makeStyles-container-9"
              >
                <div
                  class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root makeStyles-icon-10"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
                    />
                  </svg>
                </div>
                <span
                  class="makeStyles-dotContainer-11"
                >
                  <span
                    class="makeStyles-dot-12 makeStyles-dot_offline-17 "
                    title="Common:Offline"
                  />
                </span>
              </div>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Nickname 1
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                Some thoughts
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
        </ul>
      </div>
    </div>
  </div>
  <div
    data-test="sentinelEnd"
    tabindex="0"
  />
</div>
`;
